### PR TITLE
feat(traffic_light_fine_detector): add test for tlr util function

### PR DESCRIPTION
**Pull request by Shin-kyoto**
## Description

- Add test for util function in `traffic_light_fine_detector`

## Related links

[TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-6380)

## Tests performed

- I confirmed that test in this PR works with these command.

```bash
❯ colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=ON --packages-up-to traffic_light_fine_detector
❯ colcon test --packages-select traffic_light_fine_detector
Starting >>> traffic_light_fine_detector
Finished <<< traffic_light_fine_detector [2.21s]          

Summary: 1 package finished [2.75s]

❯ colcon test-result --all --verbose
build/traffic_light_fine_detector/Testing/20240529-0256/Test.xml: 5 tests, 0 errors, 0 failures, 0 skipped
build/traffic_light_fine_detector/test_results/traffic_light_fine_detector/copyright.xunit.xml: 3 tests, 0 errors, 0 failures, 0 skipped
build/traffic_light_fine_detector/test_results/traffic_light_fine_detector/cppcheck.xunit.xml: 3 tests, 0 errors, 0 failures, 3 skipped
build/traffic_light_fine_detector/test_results/traffic_light_fine_detector/lint_cmake.xunit.xml: 1 test, 0 errors, 0 failures, 0 skipped
build/traffic_light_fine_detector/test_results/traffic_light_fine_detector/traffic_light_fine_detector_tests.gtest.xml: 8 tests, 0 errors, 0 failures, 1 skipped
build/traffic_light_fine_detector/test_results/traffic_light_fine_detector/xmllint.xunit.xml: 2 tests, 0 errors, 0 failures, 0 skipped
Summary: 22 tests, 0 errors, 0 failures, 4 skipped

❯ cat build/traffic_light_fine_detector/test_results/traffic_light_fine_detector/traffic_light_fine_detector_tests.gtest.xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="8" failures="0" disabled="1" errors="0" time="0" timestamp="2024-05-29T11:56:10" name="AllTests">
  <testsuite name="CalWeightedIouTest" tests="8" failures="0" disabled="1" errors="0" time="0" timestamp="2024-05-29T11:56:10">
    <testcase name="NoOverlap" status="run" result="completed" time="0" timestamp="2024-05-29T11:56:10" classname="traffic_light_fine_detector.CalWeightedIouTest" />
    <testcase name="PartiallyOverlap1" status="run" result="completed" time="0" timestamp="2024-05-29T11:56:10" classname="traffic_light_fine_detector.CalWeightedIouTest" />
    <testcase name="PartiallyOverlap2" status="run" result="completed" time="0" timestamp="2024-05-29T11:56:10" classname="traffic_light_fine_detector.CalWeightedIouTest" />
    <testcase name="Included1" status="run" result="completed" time="0" timestamp="2024-05-29T11:56:10" classname="traffic_light_fine_detector.CalWeightedIouTest" />
    <testcase name="Included2" status="run" result="completed" time="0" timestamp="2024-05-29T11:56:10" classname="traffic_light_fine_detector.CalWeightedIouTest" />
    <testcase name="Zero" status="run" result="completed" time="0" timestamp="2024-05-29T11:56:10" classname="traffic_light_fine_detector.CalWeightedIouTest" />
    <testcase name="Negative" status="run" result="completed" time="0" timestamp="2024-05-29T11:56:10" classname="traffic_light_fine_detector.CalWeightedIouTest" />
    <testcase name="DISABLED_Uint32Max" status="notrun" result="suppressed" time="0" timestamp="1970-01-01T09:00:00" classname="traffic_light_fine_detector.CalWeightedIouTest" />
  </testsuite>
</testsuites>
```

- I checked this PR with running autoware.

## Notes for reviewers

- I added `Uint32Max` test case, and this case checks that the util function can handle `UINT32_MAX`. However, today's implementation cannot handle it, so I disabled this test case.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by Shin-kyoto** [comment]
@tzhong518 @kminoda 

Can you this PR?

**Comment by kminoda** [comment]
@tzhong518 Hi, would you review this PR? Not necessarily be a perfect review.

File: perception/traffic_light_fine_detector/test/test_nodelet.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "traffic_light_fine_detector/nodelet.hpp"
+
+#include <gtest/gtest.h>
+
+sensor_msgs::msg::RegionOfInterest createMapBasedBbox(
+  uint32_t x_offset, uint32_t y_offset, uint32_t width, uint32_t height)
```
Comment (kminoda) [inline perception/traffic_light_fine_detector/test/test_nodelet.cpp:20]: Would you add const for the arguments?
```suggestion
  const uint32_t x_offset, const uint32_t y_offset, const uint32_t width, const uint32_t height)
```

File: perception/traffic_light_fine_detector/test/test_nodelet.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  bbox.height = height;
+  bbox.score = score;
+  bbox.type = type;
+  return bbox;
+}
+
+TEST(CalWeightedIouTest, NoOverlap)
+{
+  sensor_msgs::msg::RegionOfInterest map_based_bbox = createMapBasedBbox(0, 0, 10, 10);
+  tensorrt_yolox::Object yolox_bbox = createYoloxBbox(20, 20, 10, 10, 0.9f, 0);
```
Comment (kminoda) [inline perception/traffic_light_fine_detector/test/test_nodelet.cpp:46]: Please add const
```suggestion
  const sensor_msgs::msg::RegionOfInterest map_based_bbox = createMapBasedBbox(0, 0, 10, 10);
  const tensorrt_yolox::Object yolox_bbox = createYoloxBbox(20, 20, 10, 10, 0.9f, 0);
```

**Review by kminoda — COMMENTED** [review]
Thank you for the PR :pray:

File: perception/traffic_light_fine_detector/test/test_nodelet.cpp (hunk 2 above)
Comment (Shin-kyoto) [reply perception/traffic_light_fine_detector/test/test_nodelet.cpp]: Thank you for your comment!
I modified in https://github.com/autowarefoundation/autoware.universe/pull/7075/commits/c76b5fd5a337ce0a47966628423f8e1113cbaffb

File: perception/traffic_light_fine_detector/test/test_nodelet.cpp (hunk 1 above)
Comment (Shin-kyoto) [reply perception/traffic_light_fine_detector/test/test_nodelet.cpp]: Thank you for your comment!
I modified in https://github.com/autowarefoundation/autoware.universe/pull/7075/commits/441a6ff41449bf8b12f805267e2cd1004429b0f6

File: perception/traffic_light_fine_detector/test/test_nodelet.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  sensor_msgs::msg::RegionOfInterest bbox;
+  bbox.x_offset = x_offset;
+  bbox.y_offset = y_offset;
+  bbox.width = width;
+  bbox.height = height;
+  return bbox;
+}
+
+tensorrt_yolox::Object createYoloxBbox(
+  int32_t x_offset, int32_t y_offset, int32_t width, int32_t height, float score, int32_t type)
```
Comment (tzhong518) [inline perception/traffic_light_fine_detector/test/test_nodelet.cpp]: Is const also needed here?

**Review by tzhong518 — APPROVED** [review]
LGTM

**Review by kminoda — APPROVED** [review]

**Pull request closed without merging**